lgc/builder/BuilderImpl.cpp

```diff
@@ -74,6 +74,165 @@ Value *BuilderImplBase::CreateDotProduct(Value *const vector1, Value *const vect
+  Value *result = nullptr;
+  if (hasNativeIntrinsic) {
+    int supportedN = InvalidValue;
+    int intrinsic = InvalidValue;
+    if (isDot2) {
+      intrinsic = isBothSigned ? Intrinsic::amdgcn_sdot2 : Intrinsic::amdgcn_udot2;
+      supportedN = 2;
+    } else if (isDot4) {
+      intrinsic = isBothSigned ? Intrinsic::amdgcn_sdot4 : Intrinsic::amdgcn_udot4;
+      supportedN = 4;
+    } else {
+      assert(isDot8);
+      intrinsic = isBothSigned ? Intrinsic::amdgcn_sdot8 : Intrinsic::amdgcn_udot8;
+      supportedN = 8;
+    }
+    assert(intrinsic != InvalidValue);
+    // Do null-extension
+    SmallVector<int, 8> shuffleMask;
+    for (int i = 0; i < supportedN; ++i)
+      shuffleMask.push_back(std::min(i, static_cast<int>(compCount)));
+    input1 = CreateShuffleVector(input1, Constant::getNullValue(inputTy), shuffleMask);
+    input2 = CreateShuffleVector(input2, Constant::getNullValue(inputTy), shuffleMask);
+
+    // Cast to i32 for dot4 and dot8
+    if (compBitWidth == 4 || compBitWidth == 8) {
+      input1 = CreateBitCast(input1, getInt32Ty());
+      input2 = CreateBitCast(input2, getInt32Ty());
+    }
+
+    Value *clamp = hasAccumulator ? getTrue() : getFalse();
+    accumulator = isSigned ? CreateSExt(accumulator, getInt32Ty()) : CreateZExt(accumulator, getInt32Ty());
+    result = CreateIntrinsic(intrinsic, {}, {input1, input2, accumulator, clamp}, nullptr, instName);
+  } else {
+    Value *sum = nullptr;
+    const bool canUseDot2 = isSupportCompBitwidth && isSupportSignedness && !isDot4 && !isDot8;
+    if (canUseDot2) {
+      sum = getInt32(0);
+      // Iterator over two components at a time and perform shuffle vectors and then use the intrinsic
+      unsigned intrinsic = isBothSigned ? Intrinsic::amdgcn_sdot2 : Intrinsic::amdgcn_udot2;
+      for (int compIdx = 0; compIdx < compCount; compIdx += 2) {
+        input1 = CreateShuffleVector(vector1, Constant::getNullValue(inputTy), ArrayRef<int>{compIdx, compIdx + 1});
+        input2 = CreateShuffleVector(vector2, Constant::getNullValue(inputTy), ArrayRef<int>{compIdx, compIdx + 1});
+        sum = CreateIntrinsic(intrinsic, {}, {input1, input2, sum, getFalse()}, nullptr, instName);
+      }
+    } else {
+      sum = getIntN(resBitWidth, 0);
+      for (unsigned compIdx = 0; compIdx < compCount; ++compIdx) {
+        Value *elem1 = CreateExtractElement(vector1, compIdx);
+        elem1 = isSigned ? CreateSExt(elem1, resultTy) : CreateZExt(elem1, resultTy);
+        Value *elem2 = CreateExtractElement(vector2, compIdx);
+        elem2 = isBothSigned ? CreateSExt(elem2, resultTy) : CreateZExt(elem2, resultTy);
+        sum = CreateAdd(sum, CreateMul(elem1, elem2));
+      }
+    }
+    if (hasAccumulator) {
+      if (sum->getType()->getScalarSizeInBits() < resBitWidth)
+        sum = CreateTrunc(sum, resultTy);
+      else if (sum->getType()->getScalarSizeInBits() > resBitWidth)
+        sum = isSigned ? CreateSExt(sum, resultTy) : CreateZExt(sum, resultTy);
```

Comment:
I believe the conditions here are the wrong way around. If `sum` has fewer than `resBitWidth` bits, it needs to be extended.

Reply:
I made a mistake and will fix.

Reply:
Fixed.